### ISSUE TITLE
CookieRefresh option does not refresh session age, spams logs

### ISSUE THREAD

**Issue by YenTheFirst** [oauth2-proxy/oauth2-proxy#490]
When a session is old enough to be "refreshed", its age is never reset, causing subsequent authenticated requests to continually spam "Refreshing old session cookie" in the logs.

## Expected Behavior

When:
* `--cookie-refresh=1m` is set
* multiple calls to `/oauth2/auth` are made over the course of several minutes

I expect that:
* in the logs, at most 1 request per minute has a "Refreshing old session cookie" message
* in the logs, the age of this request message never significantly exceeds my `--cookie-refresh` parameter.
* On the client, I see an updated cookie with a new `CreatedAt`

## Current Behavior

For simple CookieSession sessions backed by basic auth, every request after the initial refresh duration logs a "Refresh" message.

For OIDC sessions backed by redis, it appears to continually re-authenticate an unexpired access token, until eventually the access token expires.

## Possible Solution

When a session should be refreshed, it's `CreatedAt` could be reset to `now`. Alternatively, a second `RefreshedAt` could be tracked. Alternatively, the `--cookie-refresh` option could be removed.

## Steps to Reproduce (for bugs)

A simple reproduction of the case:

1.  Run a test oauth2_proxy (using password file auth for test simplicity)
```
docker run --mount type=bind,source=/path/to/passwordfolder,target=/data -p 127.0.0.1:4180:4180 quay.io/oauth2-proxy/oauth2-proxy  -http-address='0.0.0.0:4180' -htpasswd-file=/data/passwords.txt -cookie-secret='YmJiYmJiYmJiYmJiYmJiYg==' -client-id=foo -client-secret=bar -cookie-secure=false -cookie-refresh=60s
```
2.  log in via curl, saving the cookie
```
curl -c cookies.txt -v 'http://127.0.0.1:4180/oauth2/sign_in' --data 'rd=%2F&username=testuser&password=testpassword'
```

3.  continually re-request authentication status
```
while true; do
  date
  curl -v '127.0.0.1:4180/oauth2/auth' -c cookies.txt -b cookies.txt
  md5 cookies.txt
  tail -n1 cookies.txt | ruby -rbase64 -rjson -ne 'p JSON.parse(Base64.decode64($_.split.last.split("|").first))'
  sleep 15
done
```
4.  observe the log output:
oauth2_proxy:
```
172.17.0.1 - testuser [2020/04/13 23:33:58] [AuthSuccess] Authenticated via HtpasswdFile
172.17.0.1 - - [2020/04/13 23:33:58] 127.0.0.1:4180 POST - "/oauth2/sign_in" HTTP/1.1 "curl/7.54.0" 302 0 0.000
172.17.0.1 - testuser [2020/04/13 23:34:03] 127.0.0.1:4180 GET - "/oauth2/auth" HTTP/1.1 "curl/7.54.0" 202 0 0.000
172.17.0.1 - testuser [2020/04/13 23:34:18] 127.0.0.1:4180 GET - "/oauth2/auth" HTTP/1.1 "curl/7.54.0" 202 0 0.000
172.17.0.1 - testuser [2020/04/13 23:34:33] 127.0.0.1:4180 GET - "/oauth2/auth" HTTP/1.1 "curl/7.54.0" 202 0 0.000
172.17.0.1 - testuser [2020/04/13 23:34:49] 127.0.0.1:4180 GET - "/oauth2/auth" HTTP/1.1 "curl/7.54.0" 202 0 0.001
[2020/04/13 23:35:04] [oauthproxy.go:872] Refreshing 1m5.5872417s old session cookie for Session{email: user:testuser PreferredUsername: created:2020-04-13 23:33:58.4127583 +0000 UTC} (refresh after 1m0s)
172.17.0.1 - testuser [2020/04/13 23:35:04] 127.0.0.1:4180 GET - "/oauth2/auth" HTTP/1.1 "curl/7.54.0" 202 0 0.001
[2020/04/13 23:35:19] [oauthproxy.go:872] Refreshing 1m20.5872417s old session cookie for Session{email: user:testuser PreferredUsername: created:2020-04-13 23:33:58.4127583 +0000 UTC} (refresh after 1m0s)
172.17.0.1 - testuser [2020/04/13 23:35:19] 127.0.0.1:4180 GET - "/oauth2/auth" HTTP/1.1 "curl/7.54.0" 202 0 0.000
[2020/04/13 23:35:35] [oauthproxy.go:872] Refreshing 1m36.5872417s old session cookie for Session{email: user:testuser PreferredUsername: created:2020-04-13 23:33:58.4127583 +0000 UTC} (refresh after 1m0s)
172.17.0.1 - testuser [2020/04/13 23:35:35] 127.0.0.1:4180 GET - "/oauth2/auth" HTTP/1.1 "curl/7.54.0" 202 0 0.000
[2020/04/13 23:35:50] [oauthproxy.go:872] Refreshing 1m51.5872417s old session cookie for Session{email: user:testuser PreferredUsername: created:2020-04-13 23:33:58.4127583 +0000 UTC} (refresh after 1m0s)
172.17.0.1 - testuser [2020/04/13 23:35:50] 127.0.0.1:4180 GET - "/oauth2/auth" HTTP/1.1 "curl/7.54.0" 202 0 0.001
```
bash(edited)
```
MD5 (cookies.txt) = 19502faba8fd0ea08293130b8b577904
{"User"=>"7F31PtJ13MC1MRepdbxdR57tHIRDAHq7", "CreatedAt"=>"2020-04-13T23:33:58.4127583Z"}
...
MD5 (cookies.txt) = 7e53bd520ddb9dc1e9d7e9830fc7557c
{"User"=>"0WsAYW4zvAIKq9kvkba7hPjLg7Lsqbuh", "CreatedAt"=>"2020-04-13T23:33:58.4127583Z"}
...
MD5 (cookies.txt) = 0eb4c50b0a28563df35574bc5674a11a
{"User"=>"oP6+kfZDu6MitpHE007d0Pz+7265EWEh", "CreatedAt"=>"2020-04-13T23:33:58.4127583Z"}
...
```

## Context

While investigating other issues related to traffic and redirect loops, I was led astray by the surprising frequency of these Refresh messages, especially frequent & continual Refreshes for the same user that were well after the `--cookie-refresh` parameter. 

## Your Environment

We're using oauth2-proxy as a delegated authentication endpoint for ingress-nginx in kubernetes.
However, the same issue reproduces outside that environment.

**Comment by shamil** [comment]
I'm experiencing exactly same issue. The cookie doesn't get refreshed as it seems and this causing many unwanted calls (in our case to google: https://www.googleapis.com/oauth2/v1/tokeninfo)

**Comment by ploxiln** [comment]
If cookie-refresh does not work, it can be disabled, to avoid the spam. (In fact it is disabled by default.)

**Comment by shamil** [comment]
@ploxiln so whats the point of this param if the only ting it does is spamming logs, without even refreshing the cookie?

**Comment by ploxiln** [comment]
for correctly configured supported providers, it works. for others, you can still use oauth2-proxy without it spamming your logs.

the bug isn't the spamming of logs, the bug is that for your particular config, refresh does not work. (just having it reset the age to add a long delay to the refresh retry would be a silly fix)

**Comment by shamil** [comment]
@ploxiln can you elaborate `correctly configured supported providers`.
I use Google provider, what should I look for in configuration to see what I did wrong?

**Comment by ploxiln** [comment]
If the oauth2-proxy logs seem to indicate that refresh succeeded, then the likely problem is that updating the cookie in the browser is not working. Check your browser's debug console network tab to get more details. Are you using nginx `auth_request` mode? Configuring SetCookie header through this can be tricky. Are most of the requests from the browser "ajax" requests rather than direct page loads? I'm not sure under exactly which conditions this behavior allows updating the cookies, it's also tricky ...

**Comment by YenTheFirst** [comment]
To clarify, with my test/reproduction of the issue, I'm not going through nginx, and the issue is reproducing.

With the test, I'm talking directly talking to oauth2_proxy, and I'm getting and setting a new cookie (the `User` changes) - but the cookie has an identical `CreatedAt`

**Comment by ploxiln** [comment]
The demo in your report uses htpasswd file for initial login, which can't really be "refreshed" like other token-based auth. So my fork of the old `bitly/oauth2_proxy` it still has this logic:

```go
	if session != nil && p.CookieRefresh != time.Duration(0) && sessionAge > p.CookieRefresh && session.AccessToken != "" {
		log.Printf("%s refreshing %s old session cookie for %s (refresh after %s)", remoteAddr, sessionAge, session, p.CookieRefresh)
		saveSession = true
	}
```

(htpasswd-auth does not provide an AccessToken)

whereas in this new-upstream it has been updated to:

```go
		if session != nil {
			if session.Age() > p.CookieRefresh && p.CookieRefresh != time.Duration(0) {
				logger.Printf("Refreshing %s old session cookie for %s (refresh after %s)", session.Age(), session, p.CookieRefresh)
				saveSession = true
			}
```

so when I run your test with the fork I use, I don't see it trying to refresh at all. (When the session expires, at `-cookie-expire=168h` by default, then the user will have to htpasswd-based login again.)

So it looks like the bug now present in upstream is that it tries to refresh htpasswd-based auth cookie. This may or may not be  relevant to your original use-case and problem.

**Comment by YenTheFirst** [comment]
For my actual use case, using an OIDC provider that supports refresh, with redis session storage, I'm pretty sure "refresh" is well-defined. But the refresh logic doesn't actually attempt to refresh anything, it just re-saves the session.

**Comment by shamil** [comment]
> If the oauth2-proxy logs seem to indicate that refresh succeeded, then the likely problem is that updating the cookie in the browser is not working. Check your browser's debug console network tab to get more details. Are you using nginx `auth_request` mode? Configuring SetCookie header through this can be tricky. Are most of the requests from the browser "ajax" requests rather than direct page loads? I'm not sure under exactly which conditions this behavior allows updating the cookies, it's also tricky ...

I'm not using `auth_request` with `nginx`. In fact I'm not using `nginx` at all.
Yes most of the request are `XHR`

And `oauth2-proxy` get 200 from the provider when refreshing

```
Refreshing 50h11m8.114427678s old session cookie for Session{email:<redacted> user:<redacted> PreferredUsername: token:true id_token:true created:2020-04-13 15:22:48.885572322 +0000 UTC expires:2020-04-15 18:32:20 +0000 UTC refresh_token:true} (refresh after 1h0m0s)
[2020/04/15 17:33:57] [internal_util.go:67] 200 GET https://www.googleapis.com/oauth2/v1/tokeninfo?<redacted>
```

**Comment by JoelSpeed** [comment]
I'm not sure what the exact issue here is, but I have definitely seen this refresh behaviour working before, so I'm not sure what has changed to break this.

If anyone can provide a small, easily reproducible configuration/environment, then maybe we can narrow down the problem?

**Comment by JoelSpeed** [comment]
I've just spent a little time looking into this, I think you are facing completely separate issues

@YenTheFirst Could you share some logs, particularly the `Refreshing x old session` would be useful and any other log lines around that that look relevant. My suspicion here is that you haven't got refresh tokens in your session and so the refreshing isn't actually doing anything

@shamil You are using the google provider right? It looks like the refresh code for the google provider doesn't actually reset the `CreatedAt` time when it refreshes the session , whereas the OIDC provider does. I suspect it should be resetting that.

https://github.com/oauth2-proxy/oauth2-proxy/blob/6fa3d6c74cdf3bff1cdc6d9e0ed89e7286d8131b/providers/google.go#L260-L263

https://github.com/oauth2-proxy/oauth2-proxy/blob/6fa3d6c74cdf3bff1cdc6d9e0ed89e7286d8131b/providers/oidc.go#L127-L130

**Comment by shamil** [comment]
@JoelSpeed that's might be it

**Comment by idntfy** [comment]
I reproduced this issue with just oauth2-proxy official docker image (no nginxs etc) with the configuration which I provided in #563. In fact, right now I have two versions deployed - the old 2.2.0 on production which is good, and the new 5.0.0 on dev (didn't deploy 5.1.1 yet on live env, but locally I see the same behavior).

Inspecting logs I see that:
* the cookie refresh is made on every user request
* huge Set-Cookie header is present in response headers as well

Per request above I'm leaving here examples of logs from 5.0.0:

```
[2020/05/20 00:13:23] [oauthproxy.go:855] Refreshing 32h9m25.249941254s old session cookie for Session{email:redacted@redacted.co user:1144348redacted09441 token:true id_token:true created:2020-05-18 16:03:57.750058746 +0000 UTC expires:2020-05-20 00:38:52 +0000 UTC refresh_token:true} (refresh after 1h0m0s)
[2020/05/20 00:12:53] [oauthproxy.go:855] Refreshing 32h8m55.249941254s old session cookie for Session{email:redacted@redacted.co user:1144348redacted09441 token:true id_token:true created:2020-05-18 16:03:57.750058746 +0000 UTC expires:2020-05-20 00:38:52 +0000 UTC refresh_token:true} (refresh after 1h0m0s)
[2020/05/20 00:12:23] [oauthproxy.go:855] Refreshing 32h8m25.249941254s old session cookie for Session{email:redacted@redacted.co user:1144348redacted09441 token:true id_token:true created:2020-05-18 16:03:57.750058746 +0000 UTC expires:2020-05-20 00:38:52 +0000 UTC refresh_token:true} (refresh after 1h0m0s)
[2020/05/20 00:11:53] [oauthproxy.go:855] Refreshing 32h7m55.249941254s old session cookie for Session{email:redacted@redacted.co user:1144348redacted09441 token:true id_token:true created:2020-05-18 16:03:57.750058746 +0000 UTC expires:2020-05-20 00:38:52 +0000 UTC refresh_token:true} (refresh after 1h0m0s)
[2020/05/20 00:11:23] [oauthproxy.go:855] Refreshing 32h7m25.249941254s old session cookie for Session{email:redacted@redacted.co user:1144348redacted09441 token:true id_token:true created:2020-05-18 16:03:57.750058746 +0000 UTC expires:2020-05-20 00:38:52 +0000 UTC refresh_token:true} (refresh after 1h0m0s)
[2020/05/20 00:10:53] [oauthproxy.go:855] Refreshing 32h6m55.249941254s old session cookie for Session{email:redacted@redacted.co user:1144348redacted09441 token:true id_token:true created:2020-05-18 16:03:57.750058746 +0000 UTC expires:2020-05-20 00:38:52 +0000 UTC refresh_token:true} (refresh after 1h0m0s)
[2020/05/20 00:10:23] [oauthproxy.go:855] Refreshing 32h6m25.249941254s old session cookie for Session{email:redacted@redacted.co user:1144348redacted09441 token:true id_token:true created:2020-05-18 16:03:57.750058746 +0000 UTC expires:2020-05-20 00:38:52 +0000 UTC refresh_token:true} (refresh after 1h0m0s)
```

**Comment by JoelSpeed** [comment]
@idntfy Are your cookies being split? It's suggesting to me that the set-cookie header isn't actually being reflected into your browser (the session description looks identical, you'd expect the expires value to be updated after a refresh right?)

**Comment by idntfy** [comment]
Answering to the bot - the issue is still relevant and the behavior is well-reproducible. Have no time to provide specific answer to Joel.

**Comment by weikinhuang** [comment]
I'm also seeing this issue when I enabled this to test when setting the refresh time to `1m`. I'm using haproxy in front, `TimWolla/haproxy-auth-request`. I never see a set-cookie header in the response when logging all the auth response headers in the lua file

**Comment by Bibob7** [comment]
> I'm also seeing this issue when I enabled this to test when setting the refresh time to `1m`. I'm using haproxy in front, `TimWolla/haproxy-auth-request`. I never see a set-cookie header in the response when logging all the auth response headers in the lua file

I have exactly the same issue.

**Comment by ChrisDoernen** [comment]
We are also having this issue using the OIDC provider.

I looked into the code and found in the cookie refresh mechanism, `refreshSessionWithProvider` is called which ends up being implemented as the `RefreshSessionIfNeeded` by the provider, OIDC in our case. Like the "IfNeeded" part of the signature denotes, it prevents refreshing if the session itself has not expired - so it does nothing.

So a way to force the refresh would be necessary here. 

Does that make sense or am I missing something?

**Comment by idntfy** [comment]
Up, until it's confirmed that it is fixed I believe we can't close it. I would look myself, but the project is deprioritized as of now... Maybe some time later. We still use 2.2.0 on prod due to this issue.

**Comment by idntfy** [comment]
Up

**Comment by opteemister** [comment]
I also have that issue on my local env. Docker compose with 2 containers: quay.io/oauth2-proxy/oauth2-proxy:v7.0.1 (with oidc provider configuration - Google) + simple service (rundeck/rundeck:3.3.6)
Have same logs as provided above. On each request I have log messages about that `Refreshing 2m49.459163439s old session cookie for Session` 

@JoelSpeed please let me know if you need some additional information about it. I think I would be able to provide it. Also I can provide docker-compose.yaml file which I use on my local, so you'd be able to try.

**Comment by JoelSpeed** [comment]
What refresh period do you have configured? And just to confirm, `--provider=oidc`? Could you maybe sanitize and share some logs? Is there anything obviously erroring?

**Comment by opteemister** [comment]
Hi, @JoelSpeed . Sorry for the delayed response. 
Just to clarify. Here is my docker-compose file with all parameters set:

> version: '3'
> services:
>   oauth:
>     image: "quay.io/oauth2-proxy/oauth2-proxy"
>     ports:
>       - "80:8081"
>     links:
>       - rundeck
>     environment:
>       - "OAUTH2_PROXY_PROVIDER=oidc"
>       - "OAUTH2_PROXY_COOKIE_SAME_SITE=none"
>       - "OAUTH2_PROXY_SKIP_JWT_BEARER_TOKENS=true"
>       - "OAUTH2_PROXY_EXTRA_JWT_ISSUERS=http://service.someexternaldomain/"
>       - "OAUTH2_PROXY_OIDC_ISSUER_URL=https://accounts.google.com"
>       - "OAUTH2_PROXY_HTTP_ADDRESS=http://0.0.0.0:8081"
>       - "OAUTH2_PROXY_REDIRECT_URL=http://localhost:8080/oauth2/callback"
>       - "OAUTH2_PROXY_UPSTREAMS=http://rundeck/"
>       - "OAUTH2_PROXY_EMAIL_DOMAINS=youromdaina.com"
>       - "OAUTH2_PROXY_COOKIE_NAME=rundeck_oauth2_proxy"
>       - "OAUTH2_PROXY_CLIENT_ID="
>       - "OAUTH2_PROXY_CLIENT_SECRET="
>       - "OAUTH2_PROXY_COOKIE_SECRET="
>       - "OAUTH2_PROXY_COOKIE_DOMAIN=localhost"
>       - "OAUTH2_PROXY_COOKIE_SECURE=false"
>       - "OAUTH2_PROXY_COOKIE_EXPIRE=1h"
>       - "OAUTH2_PROXY_COOKIE_REFRESH=1m"
>       - "OAUTH2_PROXY_SKIP_AUTH_PREFLIGHT=true"
>       - "OAUTH2_PROXY_PASS_USER_HEADERS=true"
>       - "OAUTH2_PROXY_PREFER_EMAIL_TO_USER=true"
>       - "OAUTH2_PROXY_PASS_BASIC_AUTH=true"
>       - "OAUTH2_PROXY_SKIP_PROVIDER_BUTTON=false"
>       - "OAUTH2_PROXY_PASS_ACCESS_TOKEN=true"
>       - "OAUTH2_PROXY_SCOPE=profile email https://www.googleapis.com/auth/spreadsheets.readonly https://www.googleapis.com/auth/drive.file"
>   rundeck:
>     image: "rundeck/rundeck:3.3.6"
>     command:
>       - "-Dserver.port=80"
>     environment: 
>       - RUNDECK_GRAILS_URL=http://localhost:8080
>       - RUNDECK_SERVER_FORWARDED=true
>       - RUNDECK_PREAUTH_ENABLED=true
>       - RUNDECK_PREAUTH_ATTRIBUTE_NAME=REMOTE_USER_GROUPS
>       - RUNDECK_PREAUTH_DELIMITER=","
>       - RUNDECK_PREAUTH_USERNAME_HEADER=X-Forwarded-User
>       - RUNDECK_PREAUTH_ROLES_HEADER=X-Forwarded-Groups
>       - RUNDECK_PREAUTH_REDIRECT_LOGOUT=true
>       - RUNDECK_PREAUTH_REDIRECT_URL=/oauth2/sign_out

And here are logs when I just try to load one page (there are 3 requests) after 1 min after login:

First attempt:
> oauth_1    | [2021/03/21 06:49:07] [stored_session.go:122] Refreshing 10m7.853617064s old session cookie for Session{email:sanitized@sanitized.com user:101199XXXXXXX6719228 PreferredUsername: token:true id_token:true created:2021-03-21 06:38:59.146382936 +0000 UTC expires:2021-03-21 07:38:58.019597688 +0000 UTC} (refresh after 1m0s)
> rundeck_1  | [2021-03-21T06:49:07,761] INFO  web.requests "GET /user/login" 10.0.2.2 http  form 8 ? [] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36)
> oauth_1    | 10.0.2.2:51783 - sanitized@sanitized.com [2021/03/21 06:49:07] localhost:8080 GET / "/user/login" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36" 200 7637 0.204
> oauth_1    | [2021/03/21 06:49:07] [stored_session.go:122] Refreshing 10m7.853617064s old session cookie for Session{email:sanitized@sanitized.com user:101199XXXXXXX6719228 PreferredUsername: token:true id_token:true created:2021-03-21 06:38:59.146382936 +0000 UTC expires:2021-03-21 07:38:58.019597688 +0000 UTC} (refresh after 1m0s)
> oauth_1    | 10.0.2.2:51783 - sanitized@sanitized.com [2021/03/21 06:49:07] localhost:8080 GET / "/static/images/rundeck-full-logo-black.png" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36" 200 23719 0.051
> rundeck_1  | [2021-03-21T06:49:08,046] INFO  web.requests "GET /static/images/rundeck-full-logo-black.png" 10.0.2.2 http  form 20 ? [image/png] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36)
> oauth_1    | [2021/03/21 06:49:08] [stored_session.go:122] Refreshing 10m8.853617064s old session cookie for Session{email:sanitized@sanitized.com user:101199XXXXXXX6719228 PreferredUsername: token:true id_token:true created:2021-03-21 06:38:59.146382936 +0000 UTC expires:2021-03-21 07:38:58.019597688 +0000 UTC} (refresh after 1m0s)
> rundeck_1  | [2021-03-21T06:49:08,366] INFO  web.requests "GET /static/images/favicon.ico" 10.0.2.2 http  form 10 ? [image/x-icon] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36)
> oauth_1    | 10.0.2.2:51783 - sanitized@sanitized.com [2021/03/21 06:49:08] localhost:8080 GET / "/static/images/favicon.ico" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36" 200 15086 0.042
 
Second attempt:
> oauth_1    | [2021/03/21 06:49:44] [stored_session.go:122] Refreshing 10m44.853617064s old session cookie for Session{email:sanitized@sanitized.com user:101199XXXXXXX6719228 PreferredUsername: token:true id_token:true created:2021-03-21 06:38:59.146382936 +0000 UTC expires:2021-03-21 07:38:58.019597688 +0000 UTC} (refresh after 1m0s)
> rundeck_1  | [2021-03-21T06:49:44,645] INFO  web.requests "GET /user/login" 10.0.2.2 http  form 3 ? [] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36)
> oauth_1    | 10.0.2.2:51783 - sanitized@sanitized.com [2021/03/21 06:49:44] localhost:8080 GET / "/user/login" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36" 200 7637 0.192
> oauth_1    | [2021/03/21 06:49:44] [stored_session.go:122] Refreshing 10m44.853617064s old session cookie for Session{email:sanitized@sanitized.com user:101199XXXXXXX6719228 PreferredUsername: token:true id_token:true created:2021-03-21 06:38:59.146382936 +0000 UTC expires:2021-03-21 07:38:58.019597688 +0000 UTC} (refresh after 1m0s)
> rundeck_1  | [2021-03-21T06:49:45,167] INFO  web.requests "GET /static/images/rundeck-full-logo-black.png" 10.0.2.2 http  form 18 ? [image/png] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36)
> oauth_1    | 10.0.2.2:51783 - sanitized@sanitized.com [2021/03/21 06:49:44] localhost:8080 GET / "/static/images/rundeck-full-logo-black.png" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36" 200 23719 0.296
> oauth_1    | [2021/03/21 06:49:45] [stored_session.go:122] Refreshing 10m45.853617064s old session cookie for Session{email:sanitized@sanitized.com user:101199XXXXXXX6719228 PreferredUsername: token:true id_token:true created:2021-03-21 06:38:59.146382936 +0000 UTC expires:2021-03-21 07:38:58.019597688 +0000 UTC} (refresh after 1m0s)
> rundeck_1  | [2021-03-21T06:49:45,567] INFO  web.requests "GET /static/images/favicon.ico" 10.0.2.2 http  form 16 ? [image/x-icon] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36)
> oauth_1    | 10.0.2.2:51783 - sanitized@sanitized.com [2021/03/21 06:49:45] localhost:8080 GET / "/static/images/favicon.ico" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36" 200 15086 0.086

Seems no errors, but Refreshing attempt on each requests.

**Comment by JoelSpeed** [comment]
Looks like some issue with the created at being reset when the session is refreshed. Not sure what the state of this is now but @NickMeves is refactoring this in #1086 at the moment and I'm pretty sure he has this covered, so, would you be able to test this with a build of his PR?

**Comment by opteemister** [comment]
I think it is the same issue in #1086. Here are logs:

> oauth_1    | [2021/03/24 07:37:08] [stored_session.go:118] Refreshing 2m35.35260484s old session cookie for Session{email:sanitized@domain.com user:101199ZZZZZZZZZ19228 PreferredUsername: token:true id_token:true created:2021-03-24 07:34:32.64739516 +0000 UTC expires:2021-03-24 08:34:31.521322492 +0000 UTC} (refresh after 1m0s)
rundeck_1  | [2021-03-24T07:37:08,199] INFO  web.requests "GET /user/login" 10.0.2.2 http  form 15 ? [] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36)
oauth_1    | 10.0.2.2:54322 - sanitized@domain.com [2021/03/24 07:37:08] localhost:8080 GET / "/user/login" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36" 200 7637 0.111
oauth_1    | [2021/03/24 07:37:08] [stored_session.go:118] Refreshing 2m35.35260484s old session cookie for Session{email:sanitized@domain.com user:101199ZZZZZZZZZ19228 PreferredUsername: token:true id_token:true created:2021-03-24 07:34:32.64739516 +0000 UTC expires:2021-03-24 08:34:31.521322492 +0000 UTC} (refresh after 1m0s)
oauth_1    | 10.0.2.2:54322 - sanitized@domain.com [2021/03/24 07:37:08] localhost:8080 GET / "/static/images/rundeck-full-logo-black.png" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36" 200 23719 0.037
rundeck_1  | [2021-03-24T07:37:08,370] INFO  web.requests "GET /static/images/rundeck-full-logo-black.png" 10.0.2.2 http  form 18 ? [image/png] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36)
oauth_1    | [2021/03/24 07:37:08] [stored_session.go:118] Refreshing 2m35.35260484s old session cookie for Session{email:sanitized@domain.com user:101199ZZZZZZZZZ19228 PreferredUsername: token:true id_token:true created:2021-03-24 07:34:32.64739516 +0000 UTC expires:2021-03-24 08:34:31.521322492 +0000 UTC} (refresh after 1m0s)
oauth_1    | 10.0.2.2:54322 - sanitized@domain.com [2021/03/24 07:37:08] localhost:8080 GET / "/static/images/favicon.ico" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36" 200 15086 0.037
rundeck_1  | [2021-03-24T07:37:08,619] INFO  web.requests "GET /static/images/favicon.ico" 10.0.2.2 http  form 22 ? [image/x-icon] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36)

> oauth_1    | [2021/03/24 07:39:47] [stored_session.go:118] Refreshing 5m14.35260484s old session cookie for Session{email:sanitized@domain.com user:101199ZZZZZZZZZ19228 PreferredUsername: token:true id_token:true created:2021-03-24 07:34:32.64739516 +0000 UTC expires:2021-03-24 08:34:31.521322492 +0000 UTC} (refresh after 1m0s)
rundeck_1  | [2021-03-24T07:39:47,583] INFO  web.requests "GET /user/login" 10.0.2.2 http  form 11 ? [] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36)
oauth_1    | 10.0.2.2:54322 - sanitized@domain.com [2021/03/24 07:39:47] localhost:8080 GET / "/user/login" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36" 200 7637 0.165
oauth_1    | [2021/03/24 07:39:47] [stored_session.go:118] Refreshing 5m14.35260484s old session cookie for Session{email:sanitized@domain.com user:101199ZZZZZZZZZ19228 PreferredUsername: token:true id_token:true created:2021-03-24 07:34:32.64739516 +0000 UTC expires:2021-03-24 08:34:31.521322492 +0000 UTC} (refresh after 1m0s)
oauth_1    | 10.0.2.2:54322 - sanitized@domain.com [2021/03/24 07:39:47] localhost:8080 GET / "/static/images/rundeck-full-logo-black.png" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36" 200 23719 0.024
rundeck_1  | [2021-03-24T07:39:47,836] INFO  web.requests "GET /static/images/rundeck-full-logo-black.png" 10.0.2.2 http  form 6 ? [image/png] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36)
oauth_1    | [2021/03/24 07:39:48] [stored_session.go:118] Refreshing 5m15.35260484s old session cookie for Session{email:sanitized@domain.com user:101199ZZZZZZZZZ19228 PreferredUsername: token:true id_token:true created:2021-03-24 07:34:32.64739516 +0000 UTC expires:2021-03-24 08:34:31.521322492 +0000 UTC} (refresh after 1m0s)
oauth_1    | 10.0.2.2:54322 - sanitized@domain.com [2021/03/24 07:39:48] localhost:8080 GET / "/static/images/favicon.ico" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36" 200 15086 0.096
rundeck_1  | [2021-03-24T07:39:48,291] INFO  web.requests "GET /static/images/favicon.ico" 10.0.2.2 http  form 40 ? [image/x-icon] (Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36)

**Comment by geekhckr** [comment]
From the logs it looks like for each request that are sent by the browser oauth2-proxy is trying to refresh the token. Because each request sends same cookie and oauth2-proxy simply check expiry of the token and cookie refresh period and starts refreshing the cookie, it doesn't know that another request (go routine) is already in progress of doing refresh for the same cookie. 

This issue is more evident if the webpage is loading lot of assets like images or making lot of Ajax requests while loading a page.

May be this can be fixed if oauth2-proxy can check whether there is already refreshing is in progress before starting refreshing the cookie.

Because of this issue, oauth2-proxy might be making lot of request to provider as well.

**Comment by geekhckr** [comment]
I think this PR https://github.com/oauth2-proxy/oauth2-proxy/pull/1063 can fix this issue

**Comment by idntfy** [comment]
@geekhckr  does it mean that redis would be required for oauth2-proxy as a sidecar?

**Comment by geekhckr** [comment]
@idntfy that would depend on your deployment. If you are running only one instance of oauth2-proxy, then you can have a redis sidecar. But if you have multiple instances (replicas) then you need to run redis as a separate deployment.

**Comment by geekhckr** [comment]
> @geekhckr sorry, that's not what I wanted to mean, I will try to clarify:
> 
> 1. Is it _required_? having additional infrastructure components will surely hurt adoption of the tool.
Yes, from the PR# 1063, lock feature to work, you need redis.
> 2. is it _needed_? it looks like regression, and it's evident that it worked correctly before in older versions.
I have not tried the older version that you have mentioned where it was working, but without locking it is difficult to avoid the multiple requests trying to refresh the token. May be I will try to look into the old code and see how it was working.
> 
> btw, I see it could be investigated in the linked issue above - [kubeapps/kubeapps#2686](https://github.com/kubeapps/kubeapps/issues/2686) - and, probably, it's just a bug of configuration (with given configuration parameters it started behaving differently) - my team and me will try to get to confirm it in the nearest month

**Comment by JoelSpeed** [comment]
I think in the earlier versions we didn't actually have provider refresh working for many providers.

So it's possible that you were just seeing the cookie get refreshed (user accessed site, some validation, set new cookie) without OAuth2 Proxy actually checking anything with the identity provider.

We are aware that there are some issues with the refresh logic at the moment and there is some in flight work to rework how this works to make the logic more consistent across providers

@idntfy out of interest, could you confirm which provider you are using and whether you care about token refreshing or not? (Eg for oidc providers you normally have to set some scope parameter to get a refresh token)

**Comment by iliketoprogram14** [comment]
@JoelSpeed I've had the same experience as @idntfy.

I have a pretty simple setup: oauth2_proxy, nginx, and an upstream service.  I used to run oauth2_proxy v2.2 (or possibly earlier, I forget) in a non-containerized environment with Google as the auth provider.  Token refresh worked great.

I updated my setup summer of last year to docker, which included bringing oauth2_proxy up to date.  Google is still being used as the auth provider.  Token refresh has been broken ever since.  This issue is easily reproducible by setting the cookie refresh and expire parameter to low values.

From a user POV, this is a regression.  I very much care about token refreshing.

Am I understanding correctly that #1063 may fix the issue, but we have to add redis to our setup to take advantage of the fix?  Is there any way to not require redis?  I really don't want to add any more complexity to my setup.

Logs below were from reloading the page, with oauth2_proxy v7.1.3, cookie-refresh=25s, and cookie-expire=60s:

```<sanitized_ip> - fad58118-9f95-4ed4-802d-74fe41429dbe - sanitized@sanitized.com [2021/06/22 19:45:00] sanitized.com GET - "/oauth2/auth" HTTP/1.0 "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36 Edg/91.0.864.54" 202 0 0.046
[2021/06/22 19:45:00] [stored_session.go:122] Refreshing 48.675563336s old session cookie for Session{email:sanitized@sanitized.com user:<sanitized_user_id> PreferredUsername: token:true id_token:true created:2021-06-22 19:44:11.324436664 +0000 UTC expires:2021-06-22 20:44:10 +0000 UTC refresh_token:true} (refresh after 25s)
[2021/06/22 19:45:00] [stored_session.go:122] Refreshing 48.675563336s old session cookie for Session{email:sanitized@sanitized.com user:<sanitized_user_id> PreferredUsername: token:true id_token:true created:2021-06-22 19:44:11.324436664 +0000 UTC expires:2021-06-22 20:44:10 +0000 UTC refresh_token:true} (refresh after 25s)
[2021/06/22 19:45:00] [internal_util.go:69] 200 GET https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=<sanitized_token>... {
  "issued_to": "sanitized.apps.googleusercontent.com",
  "audience": "sanitized.apps.googleusercontent.com",
  "user_id": "<sanitized_user_id>",
  "scope": "https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email openid",
  "expires_in": 3550,
  "email": "sanitized@sanitized.com",
  "verified_email": true,
  "access_type": "offline"
}

<sanitized_ip> - ebcf10ee-8b75-430d-8689-b0ac6b9a0231 - sanitized@sanitized.com [2021/06/22 19:45:00] sanitized.com GET - "/oauth2/auth" HTTP/1.0 "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36 Edg/91.0.864.54" 202 0 0.048
[2021/06/22 19:45:00] [internal_util.go:69] 200 GET https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=<sanitized_token>... {
  "issued_to": "sanitized.apps.googleusercontent.com",
  "audience": "sanitized.apps.googleusercontent.com",
  "user_id": "<sanitized_user_id>",
  "scope": "https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email openid",
  "expires_in": 3550,
  "email": "sanitized@sanitized.com",
  "verified_email": true,
  "access_type": "offline"
}


<sanitized_ip> - 6b89d687-af12-455e-8a71-242994a2f5fe - sanitized@sanitized.com [2021/06/22 19:45:00] sanitized.com GET - "/oauth2/auth" HTTP/1.0 "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36 Edg/91.0.864.54" 202 0 0.043
[2021/06/22 19:45:00] [stored_session.go:122] Refreshing 48.675563336s old session cookie for Session{email:sanitized@sanitized.com user:<sanitized_user_id> PreferredUsername: token:true id_token:true created:2021-06-22 19:44:11.324436664 +0000 UTC expires:2021-06-22 20:44:10 +0000 UTC refresh_token:true} (refresh after 25s)
[2021/06/22 19:45:00] [stored_session.go:122] Refreshing 48.675563336s old session cookie for Session{email:sanitized@sanitized.com user:<sanitized_user_id> PreferredUsername: token:true id_token:true created:2021-06-22 19:44:11.324436664 +0000 UTC expires:2021-06-22 20:44:10 +0000 UTC refresh_token:true} (refresh after 25s)
[2021/06/22 19:45:00] [stored_session.go:122] Refreshing 48.675563336s old session cookie for Session{email:sanitized@sanitized.com user:<sanitized_user_id> PreferredUsername: token:true id_token:true created:2021-06-22 19:44:11.324436664 +0000 UTC expires:2021-06-22 20:44:10 +0000 UTC refresh_token:true} (refresh after 25s)
[2021/06/22 19:45:00] [stored_session.go:122] Refreshing 48.675563336s old session cookie for Session{email:sanitized@sanitized.com user:<sanitized_user_id> PreferredUsername: token:true id_token:true created:2021-06-22 19:44:11.324436664 +0000 UTC expires:2021-06-22 20:44:10 +0000 UTC refresh_token:true} (refresh after 25s)
[2021/06/22 19:45:00] [internal_util.go:69] 200 GET https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=<sanitized_token>... {
  "issued_to": "sanitized.apps.googleusercontent.com",
  "audience": "sanitized.apps.googleusercontent.com",
  "user_id": "<sanitized_user_id>",
  "scope": "https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email openid",
  "expires_in": 3550,
  "email": "sanitized@sanitized.com",
  "verified_email": true,
  "access_type": "offline"
}```

**Comment by opteemister** [comment]
Raising up

**Comment by NickMeves** [comment]
Is this not fixed in the new `RefreshSession` implementation where we will refresh at the cadence you specify regardless of if the access/id tokens are still valid or not?

It doesn't do dynamic timing though, you will need to configure your refresh cadence manually knowing your access/ID token TTLs.

**Comment by Xwaffle1** [comment]
Has anyone confirmed this working? As of oauth2-proxy:v7.1.3 I believe sessions are not refreshed correctly using  oidc

When I load or refresh my own web app it does ~7 API calls through a Pod on K8s, and each call triggers a message: 

```
[2021/09/13 14:20:58] [stored_session.go:122] Refreshing 4m41.771150023s old session cookie for 
Session{email:staff@company.com user:ca2f3433-6f96-4354-8194-e77de6313c29 PreferredUsername:staff token:true 
id_token:true created:2021-09-13 14:16:16.228849977 +0000 UTC expires:2021-09-13 14:21:16.226134643 +0000 UTC 
refresh_token:true groups:[Staff]} (refresh after 2m0s)
```

This was setup with a 5 minute expiration, and 2 minute refresh for testing.

**Comment by JoelSpeed** [comment]
I think the fixes that Nick is mentioning haven't made it to a release just yet, have you got the ability to test this out using our latest master branch?

**Comment by idntfy** [comment]
We've been waiting and not updating the proxy from 2.2.0 hoping it will get fixed. In 2.2.0 this bug didn't exist.

Now I'm testing the version 7.3.0, the latest one, the problem still persists.

<img width="1079" alt="image" src="https://user-images.githubusercontent.com/3023928/194957032-6634517e-25ba-4898-8b20-cda7812c63de.png">

_“What Is It They’ve Sent Us?” “Hope.”_

@JoelSpeed do we have hope? Thanks

**Comment by JoelSpeed** [comment]
Hey, please bear in mind that this project is a passion project. It has no commercial sponsorship and relies on community input, by those who use the project, to get things fixed.

If you have the time to work out where the bug is, and raise a PR to fix it, then the maintainers can help get that merged

**Comment by attiss** [comment]
Hi folks,

I was experimenting with `cookie-refresh` and bumped into this issue (using 7.4.0 and an OIDC provider).

```
[2023/03/04 22:30:15] [stored_session.go:189] Refreshing session - User: redacted; SessionAge: 1m34.866260815s
172.30.12.69:29457 - 4846a6e8dd05b51f6aa9ce068ebb45bd - redacted@redacted [2023/03/04 22:30:15] oauth2-proxy.default.svc.cluster.local GET - "/oauth2-proxy/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.004

[2023/03/04 22:30:20] [stored_session.go:189] Refreshing session - User: redacted; SessionAge: 1m39.866260815s
172.30.12.69:11133 - 04851a299b47df5fb19379s23fee54fa - redacted@redacted [2023/03/04 22:30:20] oauth2-proxy.default.svc.cluster.local GET - "/oauth2-proxy/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.001

[2023/03/04 22:30:23] [stored_session.go:189] Refreshing session - User: redacted; SessionAge: 1m42.866260815s
172.30.12.69:11145 - 12f56c411d5d7ee90847dcdbe054d3f5 - redacted@redacted [2023/03/04 22:30:23] oauth2-proxy.default.svc.cluster.local GET - "/oauth2-proxy/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.002

[2023/03/04 22:30:26] [stored_session.go:189] Refreshing session - User: redacted; SessionAge: 1m45.866260815s
172.30.12.69:48879 - bdeaeab2c72ca7975c8c22604eeb4397 - redacted@redacted [2023/03/04 22:30:26] oauth2-proxy.default.svc.cluster.local GET - "/oauth2-proxy/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.002

[2023/03/04 22:30:30] [stored_session.go:189] Refreshing session - User: redacted; SessionAge: 1m49.866260815s
172.30.12.69:48891 - d5629c6352452081584eeda0f56b2306 - redacted@redacted [2023/03/04 22:30:30] oauth2-proxy.default.svc.cluster.local GET - "/oauth2-proxy/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.004
```

Could we please have this issue reopened?

**Comment by JoelSpeed** [comment]
Seems there's a bug here as there's been a few reports, does anyone have time to look for it and raise a PR?

**Comment by attiss** [comment]
(@JoelSpeed thanks for reopening the issue. I bumped into some configuration issues on my side that caused this problem. However since I resolved these configuration issues, cookie-refresh works fine for me using 7.4.0 and an OIDC provider.)

**Comment by patrickjahns** [comment]
@attiss  
Can you share the configuration issues? I am currently debugging this with okta and facing a similar issue as you described

**Comment by patrickjahns** [comment]
Thank you for the feedback - I managed to trace it down to missing scope `offline_access` - with that scope the refresh token will be properly issued from okta.

**Comment by marcellodesales** [comment]
Following up with what @patrickjahns described above:

# 🔧  Offline support for OKTA

* The solution is with the following parameters
* DO NOT use double-quotes
  * `--scope=openid profile email offline_access`

```yaml
      - name: oauth2-proxy
        image: "quay.io/oauth2-proxy/oauth2-proxy:v7.4.0"
        imagePullPolicy: IfNotPresent
        args:
          - --http-address=0.0.0.0:4180
          - --https-address=0.0.0.0:4443
          - --metrics-address=0.0.0.0:44180
          - --auth-logging=true
          - --code-challenge-method=S256
          - --cookie-domain=chat.gpt.company.com
          - --cookie-expire=12h
          - --cookie-name=_oauth2_proxy_gpt_prd_company_com
          - --cookie-refresh=1h
          # Someone reported that for OKTA we need to provide the offline_access https://github.com/oauth2-proxy/oauth2-proxy/issues/490#issuecomment-1576644398
          # The way to support scope https://github.com/oauth2-proxy/oauth2-proxy/issues/793#issuecomment-748952258
          # OKTA team requires these 3 for offline_access https://company.slack.com/archives/CL2L1D336/p1667594219319339?thread_ts=1667592264.463819&cid=CL2L1D336
          # https://developer.okta.com/docs/reference/api/oidc/#limits
          - --scope=openid profile email offline_access
          - ...
          - ...
          - ...
```

# 🔊 Logs

* The authentication will now work, providing the token for OKTA properly and setting the value in the auth values

```console
127.0.0.6:42875 - be6f1faf-8e46-4b2a-bc10-6b13629be804 - Marcello.DeSales@company.com [2023/07/08 03:32:54] [AuthSuccess] Authenticated via OAuth2: Session{email:Marcello.DeSales@company.com user:00uwhm2r9o5C7Q8Gz696 PreferredUsername:Marcello.DeSales@company.com token:true id_token:true created:2023-07-08 03:32:54.621711875 +0000 UTC m=+40.282354673 expires:2023-07-08 15:32:54.19472548 +0000 UTC m=+43239.855368274 refresh_token:true}
```

# 🐛 Errors

> **NOTE**: If this is used with `--scope="openid profile email offline_access"`, the following error occurs

A lot of docs have the value with double-quotes...

```console
[2023/07/08 03:31:35] [oauthproxy.go:807] Error while parsing OAuth2 callback: invalid_scope
{ "client":"127.0.0.6:35495", "method":"GET", "username":"-", "timestamp":"2023/07/08 03:31:35", "host":"chat.gpt.company.com", "upstream":"-", "url":"/oauth2/callback?state=IboSMafzJDD1pWQGhBDkD_hVkBnAJVl30RL0FlKFIg8%3A%2F&error=invalid_scope&error_description=Requests+for+ID+tokens+or+access+tokens+with+OpenID+scopes+require+the+%27openid%27+scope.", "protocol":"HTTP/1.1", "user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36", "status":"403", "length":"2811", "response-time":"0.000" }
```

**Comment by mbastress** [comment]
We were using an Istio AuthorizationPolicy with CUSTOM action and extension provider as described [here](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/). We were missing the following:

`headersToDownstreamOnAllow: ["set-cookie"]`

Adding that fixed the issue.

**Comment by redboxstudio** [comment]
> Hi folks,
> 
> I was experimenting with `cookie-refresh` and bumped into this issue (using 7.4.0 and an OIDC provider).
> 
> ```
> [2023/03/04 22:30:15] [stored_session.go:189] Refreshing session - User: redacted; SessionAge: 1m34.866260815s
> 172.30.12.69:29457 - 4846a6e8dd05b51f6aa9ce068ebb45bd - redacted@redacted [2023/03/04 22:30:15] oauth2-proxy.default.svc.cluster.local GET - "/oauth2-proxy/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.004
> 
> [2023/03/04 22:30:20] [stored_session.go:189] Refreshing session - User: redacted; SessionAge: 1m39.866260815s
> 172.30.12.69:11133 - 04851a299b47df5fb19379s23fee54fa - redacted@redacted [2023/03/04 22:30:20] oauth2-proxy.default.svc.cluster.local GET - "/oauth2-proxy/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.001
> 
> [2023/03/04 22:30:23] [stored_session.go:189] Refreshing session - User: redacted; SessionAge: 1m42.866260815s
> 172.30.12.69:11145 - 12f56c411d5d7ee90847dcdbe054d3f5 - redacted@redacted [2023/03/04 22:30:23] oauth2-proxy.default.svc.cluster.local GET - "/oauth2-proxy/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.002
> 
> [2023/03/04 22:30:26] [stored_session.go:189] Refreshing session - User: redacted; SessionAge: 1m45.866260815s
> 172.30.12.69:48879 - bdeaeab2c72ca7975c8c22604eeb4397 - redacted@redacted [2023/03/04 22:30:26] oauth2-proxy.default.svc.cluster.local GET - "/oauth2-proxy/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.002
> 
> [2023/03/04 22:30:30] [stored_session.go:189] Refreshing session - User: redacted; SessionAge: 1m49.866260815s
> 172.30.12.69:48891 - d5629c6352452081584eeda0f56b2306 - redacted@redacted [2023/03/04 22:30:30] oauth2-proxy.default.svc.cluster.local GET - "/oauth2-proxy/auth" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36" 202 0 0.004
> ```
> 
> ~Could we please have this issue reopened?~
> 
> Update: turned out that I had a configuration issue on the OIDC side and the refresh tokens were not passed to _oauth2-proxy_ alongside the access token & ID token. Once the configuration issue was resolved, the session refresh started working.
> 
> ~(However I bumped into another issue: #1948; after the session is successfully renewed and the cookies are updated, the next validation of the cookies fail.)~
> 
> Update 2: turned out that I had yet another configuration issue with our reverse proxy setup and our cookies were not updated correctly. Since these configuration issues are fixed, **cookie-refresh works just fine**!

hi, could you share what exactly was done to normalize the work? I seem to be experiencing similar problems. At the first login, I receive the oauth2-proxy log as authsuccess and see in it the current creation time and the expiration time equal to the one specified in the keycloak access token lifespan. according to the specified refresh cookie, the logs are present, but as soon as the access token expires, my application begins to receive 302 responses before the page is refreshed, which violates the principle of operation of a single-page application

**Comment by attiss** [comment]
@redboxstudio
- Issue 1: OIDC provider did not include refresh token in the response
    - Fix: enable refresh tokens: https://github.com/oauth2-proxy/oauth2-proxy/issues/490#issuecomment-1568772626
- Issue 2: cookie update problem
    - Official documentation: https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview#configuring-for-use-with-the-nginx-auth_request-directive
    - TL;DR: when using cookies for session storage oauth2-proxy splits cookies that are larger than 4 kB. This might be an issue when you are using nginx as a reverse proxy for your setup, because the `auth_request` module does not handle passing more than one cookie back to the client. When a refresh happens, only the first part of the cookie gets updated, but the second (or third, and so on) don't. After this, oauth2-proxy won't be able to decode decode the session information stored in the client browser and will require a new login.
    - Solution is to add a lua script to pass all the cookies back to the client, not just the first one. (This requires lua support in your nginx setup.) See the example in the official documentation.

**Comment by kazaroff** [comment]
Had the same issue. Solution was to run oauth2-proxy with redis storage for session data (avoiding multiple cookies).
